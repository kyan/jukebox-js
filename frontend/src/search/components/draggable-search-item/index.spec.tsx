import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import DraggableSearchItem from './index'

const validTrack = {
  name: 'track name',
  artist: {
    name: 'artist name'
  },
  album: {
    name: 'album name'
  },
  metrics: {},
  image: 'image1',
  explicit: false
}

describe('DraggableSearchItem', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  describe('render', () => {
    const onRemoveMock = jest.fn().mockName('onRemoveMock')
    const swapMock = jest.fn().mockName('swapMock')

    test('with valid props it renders correctly', () => {
      const { asFragment } = render(
        <DraggableSearchItem i={1} action={swapMock} track={validTrack} onRemove={onRemoveMock} />
      )
      expect(asFragment().firstChild).toMatchSnapshot()
    })

    test('clicking remove link removes track from list', () => {
      const { getByText } = render(
        <DraggableSearchItem i={1} action={swapMock} track={validTrack} onRemove={onRemoveMock} />
      )
      fireEvent.click(getByText('Remove'))
      expect(onRemoveMock).toHaveBeenCalledTimes(1)
    })

    test('onDrop', () => {
      const getDataMock = jest.fn().mockReturnValue('20')
      const { getByTitle } = render(
        <DraggableSearchItem i={1} action={swapMock} track={validTrack} onRemove={onRemoveMock} />
      )
      fireEvent.drop(getByTitle('You can drag this to sort.'), {
        dataTransfer: {
          getData: getDataMock
        }
      })
      expect(getDataMock).toHaveBeenCalledWith('text/plain')
      expect(swapMock).toHaveBeenCalledWith(1, 20)
    })

    test('onDragStart', () => {
      const setDataMock = jest.fn()
      const { getByTitle } = render(
        <DraggableSearchItem i={1} action={swapMock} track={validTrack} onRemove={onRemoveMock} />
      )
      fireEvent.dragStart(getByTitle('You can drag this to sort.'), {
        dataTransfer: {
          setData: setDataMock
        }
      })
      expect(setDataMock).toHaveBeenCalledWith('text/plain', '1')
    })
  })
})
