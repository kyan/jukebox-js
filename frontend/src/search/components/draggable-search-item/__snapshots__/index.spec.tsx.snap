// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraggableSearchItem render with valid props it renders correctly 1`] = `
<DraggableSearchItem
  action={[MockFunction swapMock]}
  i={1}
  onClick={[MockFunction onClickMock]}
  onRemove={[MockFunction onRemoveMock]}
  track={
    Object {
      "album": Object {
        "name": "album name",
      },
      "artist": Object {
        "name": "artist name",
      },
      "explicit": false,
      "image": "image1",
      "metrics": Object {},
      "name": "track name",
    }
  }
>
  <div
    className="draggable search-list-item-draggable"
    draggable={true}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
    title="You can drag this to sort."
  >
    <SearchItem
      track={
        Object {
          "album": Object {
            "name": "album name",
          },
          "artist": Object {
            "name": "artist name",
          },
          "explicit": false,
          "image": "image1",
          "metrics": Object {},
          "name": "track name",
        }
      }
    >
      <div
        className="search-list-item"
      >
        <Image
          alt="track name"
          as="img"
          className="search-list-item__image"
          disabled={false}
          floated="left"
          size="tiny"
          src="image1"
          title="Click to add - track name - artist name"
          ui={true}
        >
          <img
            alt="track name"
            className="ui tiny left floated image search-list-item__image"
            src="image1"
            title="Click to add - track name - artist name"
          />
        </Image>
        <ListContent>
          <div
            className="content"
          >
            <div
              className="search-list-item__header"
            >
              track name
               - 
              artist name
            </div>
            <div
              className="search-list-item__content"
            >
              album name
            </div>
            <VoteInfo
              metrics={Object {}}
            >
              <VotedBy
                show={false}
                size="mini"
              />
            </VoteInfo>
          </div>
        </ListContent>
      </div>
    </SearchItem>
    <span
      className="search-list-item__remove"
      onClick={[MockFunction onRemoveMock]}
    >
      Remove
    </span>
  </div>
</DraggableSearchItem>
`;
